### PR TITLE
Introduce a prototype of a "header guard enforcement" tool

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/133415.

---

This is a prototype I threw together in about 1-2 hours. It enforces and automatically fixes header guards that don't match the [the Google C++ style guide](https://google.github.io/styleguide/cppguide.html#The__define_Guard). For example, here is (trimmed) output at HEAD:

```txt
line 5, column 1 of impeller/aiks/picture.h: Unexpected #pragma once
  ╷
5 │ #pragma once
  │ ^^^^^^^^^^^^
  ╵
line 5, column 1 of flow/stopwatch.h: Expected #ifndef FLUTTER_FLOW_STOPWATCH_H_
  ╷
5 │ #ifndef FLUTTER_FLOW_INSTRUMENTATION_H_
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
The following 731 files have invalid header guards:
```

### CODE REVIEW

File: tools/header_guard_check/lib/header_guard_check.dart [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      }
+      if (!excluded) {
+        files.add(entity);
+      }
+    }
+  }
+
+    // Check each file.
+    final List<HeaderFile> badFiles = <HeaderFile>[];
+    for (final io.File file in files) {
```
Comment (gaaclarke) [inline tools/header_guard_check/lib/header_guard_check.dart:62]: Can we use my package to easily parallelize this? https://pub.dev/packages/pmap

Comment (matanlurey) [reply tools/header_guard_check/lib/header_guard_check.dart:62]: Theoretically, though I admit for this example the cost of going async is higher than scanning the text file. I'd be open to it when this takes longer than 1-2s (it currently takes about 1s).

File: tools/header_guard_check/lib/header_guard_check.dart [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  final Engine source;
+
+  /// Path directories to exclude from the check.
+  final List<String> exclude;
+
+  /// Runs the header guard check.
+  Future<int> run() async {
+    final List<io.File> files = <io.File>[];
+
+  // Recursive search for header files.
```
Comment (zanderso) [inline tools/header_guard_check/lib/header_guard_check.dart:42]: Is it my eyes, or is the indentation a bit off from here down?

File: tools/header_guard_check/pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (zanderso) [inline tools/header_guard_check/pubspec.yaml:1]: This package should be listed in https://github.com/flutter/engine/blob/main/tools/pub_get_offline.py#L22

File: tools/header_guard_check/lib/header_guard_check.dart (hunk 2 above)
Comment (matanlurey) [reply tools/header_guard_check/lib/header_guard_check.dart:42]: Yeah it's not aligned, thanks for catching :)

File: tools/header_guard_check/pubspec.yaml (hunk 3 above)
Comment (matanlurey) [reply tools/header_guard_check/pubspec.yaml:1]: Ack. I was just getting feedback on whether to continue this work mostly :)

**Comment by Hixie** [comment]
test-exempt: tested by https://github.com/flutter/engine/pull/48903

**Comment by matanlurey** [comment]
Ok, I think I'm ready to land the tool (without enforcing it on CI - that will be another PR). PTAL.

**Review by gaaclarke — APPROVED** [review]
LGTM, just a couple of suggestions.  There were a few places where there may be false positives, but we'll get those sorted out easy if this is running on ci.

**Review by zanderso — COMMENTED** [review]
Needs an entry in https://github.com/flutter/engine/blob/main/testing/run_tests.py#L975 to run the tests on CI.

File: tools/header_guard_check/lib/header_guard_check.dart [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  /// Path directories to exclude from the check.
+  final List<String> exclude;
+
+  /// Runs the header guard check.
+  Future<int> run() async {
+    final List<io.File> files = <io.File>[];
+
+    // Recursive search for header files.
+    final io.Directory dir = source.flutterDir;
+    await for (final io.FileSystemEntity entity in dir.list(recursive: true)) {
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/header_guard_check/lib/header_guard_check.dart]: Done! Or I misunderstood your comment, so please feel free to add more.

File: tools/header_guard_check/lib/header_guard_check.dart [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        }
+        if (!excluded) {
+          files.add(entity);
+        }
+      }
+    }
+
+    // Check each file.
+    final List<HeaderFile> badFiles = <HeaderFile>[];
+    for (final io.File file in files) {
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/header_guard_check/lib/header_guard_check.dart]: Done!

File: tools/header_guard_check/lib/src/header_file.dart [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+  ///
+  /// This is `null` if the file does not have a `#pragma once` directive.
+  final SourceSpan? pragmaOnce;
+
+  static final RegExp _nonAlphaNumeric = RegExp(r'[^a-zA-Z0-9]');
+
+  /// Returns the expected header guard for this file, relative to [engineRoot].
+  ///
+  /// For example, if the file is `foo/bar/baz.h`, this will return `FLUTTER_FOO_BAR_BAZ_H_`.
+  String expectedName({required String engineRoot}) {
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/header_guard_check/lib/src/header_file.dart]: Done!

File: tools/header_guard_check/lib/src/header_file.dart [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    final String relativePath = p.relative(path, from: engineRoot);
+    final String underscoredRelativePath = p.withoutExtension(relativePath).replaceAll(_nonAlphaNumeric, '_');
+    return 'FLUTTER_${underscoredRelativePath.toUpperCase()}_H_';
+  }
+
+  /// Updates the file at [path] to have the expected header guard.
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/header_guard_check/lib/src/header_file.dart:136]: Added a comment.

**Comment by matanlurey** [comment]
> Needs an entry in https://github.com/flutter/engine/blob/main/testing/run_tests.py#L975 to run the tests on CI.

Done.

File: tools/header_guard_check/lib/header_guard_check.dart [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  final bool fix;
+
+  /// Path directories to include in the check.
+  final List<String> include;
+
+  /// Path directories to exclude from the check.
+  final List<String> exclude;
+
+  /// Runs the header guard check.
+  Future<int> run() async {
```
Comment (gaaclarke) [inline tools/header_guard_check/lib/header_guard_check.dart:49]: 👍

**Pull request merged**